posix/basic-unix.lisp

```diff
@@ -223,3 +223,18 @@
   "Create a directory."
   (path filename-designator)
   (mode mode))
+
+(defsyscall ("realpath" %realpath) :pointer
+  "Returns the canonicalized absolute pathname"  
+  (path filename-designator)
+  (resolved-name :pointer))
+
+(defun realpath (path)
+  "Returns the canonicalized absolute pathname"
+  (with-foreign-pointer (resolved-name path-max)
+    (let ((result (%realpath path resolved-name)))
+      (when (not (null-pointer-p result))
+        (multiple-value-bind (string length)
+            (foreign-string-to-lisp resolved-name)
+          (declare (ignore length))
+          string)))))
```

Comment:
`foreign-string-to-lisp` returns `nil` when passed a null pointer so you could simplify this to

```lisp
(with-foreign-pointer (resolved-name path-max)
  (foreign-string-to-lisp (%realpath path resolved-name)))
```

Or you could specify `%realpath` to have a `:string` return type and remove the call to `foreign-string-to-lisp`, I think.